api/service/stagedsync/stagedsync.go

```diff
@@ -0,0 +1,1321 @@
+	return &RevertState{id, revertPoint, currentProgress, common.Hash{}, s}
+}
+
+func (s *StagedSync) CleanUpStageState(id SyncStageID, forwardProgress uint64, tx kv.Tx, db kv.RwDB) (*CleanUpState, error) {
+	var pruneProgress uint64
+	var err error
+
+	if errV := CreateView(context.Background(), db, tx, func(tx kv.Tx) error {
+		pruneProgress, err = GetStageCleanUpProgress(tx, id, s.isBeacon)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); errV != nil {
+		return nil, errV
+	}
+
+	return &CleanUpState{id, forwardProgress, pruneProgress, s}, nil
+}
+
+func (s *StagedSync) NextStage() {
+	if s == nil {
+		return
+	}
+	s.currentStage++
+}
+
+// IsBefore returns true if stage1 goes before stage2 in staged sync
+func (s *StagedSync) IsBefore(stage1, stage2 SyncStageID) bool {
+	idx1 := -1
+	idx2 := -1
+	for i, stage := range s.stages {
+		if stage.ID == stage1 {
+			idx1 = i
+		}
+
+		if stage.ID == stage2 {
+			idx2 = i
+		}
+	}
+
+	return idx1 < idx2
+}
+
+// IsAfter returns true if stage1 goes after stage2 in staged sync
+func (s *StagedSync) IsAfter(stage1, stage2 SyncStageID) bool {
+	idx1 := -1
+	idx2 := -1
+	for i, stage := range s.stages {
+		if stage.ID == stage1 {
+			idx1 = i
+		}
+
+		if stage.ID == stage2 {
+			idx2 = i
+		}
+	}
+
+	return idx1 > idx2
+}
+
+// RevertTo reverts the stage to a specific height
+func (s *StagedSync) RevertTo(revertPoint uint64, invalidBlock common.Hash) {
+	utils.Logger().Info().
+		Interface("invalidBlock", invalidBlock).
+		Uint64("revertPoint", revertPoint).
+		Msgf("[STAGED_SYNC] Reverting blocks")
+	s.revertPoint = &revertPoint
+	s.invalidBlock = invalidBlock
+}
+
+func (s *StagedSync) Done() {
+	s.currentStage = uint(len(s.stages))
+	s.revertPoint = nil
+}
+
+func (s *StagedSync) IsDone() bool {
+	return s.currentStage >= uint(len(s.stages)) && s.revertPoint == nil
+}
+
+func (s *StagedSync) LogPrefix() string {
+	if s == nil {
+		return ""
+	}
+	return s.logPrefixes[s.currentStage]
+}
+
+func (s *StagedSync) SetCurrentStage(id SyncStageID) error {
+	for i, stage := range s.stages {
+		if stage.ID == id {
+			s.currentStage = uint(i)
+			return nil
+		}
+	}
+	utils.Logger().Error().
+		Interface("stage id", id).
+		Msgf("[STAGED_SYNC] stage not found")
+
+	return ErrStageNotFound
+}
+
+func New(ctx context.Context,
+	ip string,
+	port string,
+	peerHash [20]byte,
+	bc core.BlockChain,
+	role nodeconfig.Role,
+	isBeacon bool,
+	isExplorer bool,
+	db kv.RwDB,
+	stagesList []*Stage,
+	revertOrder RevertOrder,
+	pruneOrder CleanUpOrder,
+	TurboMode bool,
+	UseMemDB bool,
+	doubleCheckBlockHashes bool,
+	maxBlocksPerCycle uint64,
+	maxBackgroundBlocks uint64,
+	maxMemSyncCycleSize uint64,
+	verifyAllSig bool,
+	verifyHeaderBatchSize uint64,
+	insertChainBatchSize int,
+	logProgress bool) *StagedSync {
+
+	revertStages := make([]*Stage, len(stagesList))
+	for i, stageIndex := range revertOrder {
+		for _, s := range stagesList {
+			if s.ID == stageIndex {
```

Comment:
If this condition is false, revertStages[i] will be nil, and it will panic somewhere, but not here. 
This should be checked, otherwise it leads to tricky bugs.

Reply:
It always is true for one of the items, because each item is one of the stages. Basically, second for loop is to find the stage in stageList. We never have an stage in revertStages which is not in stageList.